core/trino-main/src/main/java/io/trino/execution/scheduler/UniformNodeSelector.java

```diff
@@ -202,6 +213,14 @@ else if (!splitWaitingForAnyNode) {
                     blockedExactNodes.addAll(candidateNodes);
                 }
```

Comment:
actually, I don't think you can exit loop earlier (in this form), see the code above:
```
            else {
                if (split.isRemotelyAccessible()) {
                    splitWaitingForAnyNode = true;
                }
                // Exact node set won't matter, if a split is waiting for any node
                else if (!splitWaitingForAnyNode) {
                    blockedExactNodes.addAll(candidateNodes);
                }
     } 
```

`blockedExactNodes` will be populated with nodes that could unblock split processing. If you quit too early, then you won't wait for such nodes to unblock.